hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMDirectoryCreateRequest.java

```diff
@@ -0,0 +1,282 @@
+import org.apache.hadoop.ozone.security.acl.IAccessAuthorizer;
+import org.apache.hadoop.ozone.security.acl.OzoneObj;
+import org.apache.hadoop.util.Time;
+import org.apache.hadoop.utils.db.cache.CacheKey;
+import org.apache.hadoop.utils.db.cache.CacheValue;
+
+
+import static org.apache.hadoop.ozone.om.exceptions.OMException.ResultCodes.BUCKET_NOT_FOUND;
+import static org.apache.hadoop.ozone.om.exceptions.OMException.ResultCodes.FILE_ALREADY_EXISTS;
+import static  org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.BUCKET_LOCK;
+/**
+ * Handle create directory request.
+ */
+public class OMDirectoryCreateRequest extends OMClientRequest
+    implements OMKeyRequest {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMDirectoryCreateRequest.class);
+
+  public OMDirectoryCreateRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) {
+    CreateDirectoryRequest createDirectoryRequest =
+        getOmRequest().getCreateDirectoryRequest();
+    Preconditions.checkNotNull(createDirectoryRequest);
+
+    KeyArgs.Builder newKeyArgs = createDirectoryRequest.getKeyArgs()
+        .toBuilder().setModificationTime(Time.now());
+
+    CreateDirectoryRequest.Builder newCreateDirectoryRequest =
+        createDirectoryRequest.toBuilder().setKeyArgs(newKeyArgs);
+
+    return getOmRequest().toBuilder().setCreateDirectoryRequest(
+        newCreateDirectoryRequest).build();
+
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long transactionLogIndex) {
+
+    KeyArgs keyArgs = getOmRequest().getCreateDirectoryRequest().getKeyArgs();
+
+    String volumeName = keyArgs.getVolumeName();
+    String bucketName = keyArgs.getBucketName();
+    String keyName = keyArgs.getKeyName();
+
+    OMResponse.Builder omResponse =
+        OzoneManagerProtocolProtos.OMResponse.newBuilder().setCmdType(
+            OzoneManagerProtocolProtos.Type.CreateDirectory).setStatus(
+            OzoneManagerProtocolProtos.Status.OK);
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumCreateDirectory();
+
+    AuditLogger auditLogger = ozoneManager.getAuditLogger();
+    OzoneManagerProtocolProtos.UserInfo userInfo = getOmRequest().getUserInfo();
+
+    Map<String, String> auditMap = buildKeyArgsAuditMap(keyArgs);
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    boolean acquiredLock = false;
+    IOException exception = null;
+    OmKeyInfo dirKeyInfo = null;
+
+    try {
+      // check Acl
+      if (ozoneManager.getAclsEnabled()) {
+        checkAcls(ozoneManager, OzoneObj.ResourceType.BUCKET,
+            OzoneObj.StoreType.OZONE, IAccessAuthorizer.ACLType.WRITE,
+            volumeName, bucketName, keyName);
+      }
+
+      // Check if this is the root of the filesystem.
+      if (keyName.length() == 0) {
+        return new OMDirectoryCreateResponse(null,
```

Comment:
A future improvement unrelated to your patch - replace null with Optional. Or at least add \@Nullable annotation on the parameter.

Reply:
Added @nullable. Will open a new jira to do the same for Other OMResponse classes.

Reply:
https://issues.apache.org/jira/browse/HDDS-1738